### DISCUSSION TITLE
성능 향상을 위한 키 캐시(MyISAM 스토리지 엔진) 기능이란?

### DISCUSSION THREAD

**Discussion by CODe5753** [Well-Founded-Confidence/RealMySQL#5] [Q&A] [answered]
InnoDB 스토리지 엔진에서는 없는 MyISAM 스토리지 엔진에서만 지원되는 기능이라는데 이게 뭘까요

**Comment by menthamin** [comment]
- MyISAM 키 캐시: 인덱스를 캐싱하는 기능
(마이아이삼)

**Comment by Soyflat** [comment, marked as answer]
키 캐싱(Key Caching)과 버퍼 풀(Buffer Pool)은 모두 데이터베이스 성능을 향상시키는 데 사용되는 기술이지만, 각각 다른 목적과 방식으로 동작합니다.

`키 캐싱: 키 캐싱은 데이터베이스의 인덱스를 더 빠르게 조회`하기 위해 사용됩니다. 인덱스를 참조하는 쿼리는 해당 인덱스의 B-트리를 탐색해야 합니다. 이때 B-트리의 노드를 디스크에서 읽어와야 하므로, 디스크 I/O가 발생할 수 있습니다. 하지만 키 캐싱을 사용하면 자주 사용되는 인덱스의 B-트리 노드를 메모리에 캐시하여 디스크 I/O를 줄일 수 있습니다. `MySQL에서는 MyISAM 스토리지 엔진에서 키 캐싱을 사용할 수 있습니다. MyISAM 테이블의 인덱스는 key_buffer_size 설정 값에 의해 캐시됩니다.`

`버퍼 풀: 버퍼 풀은 디스크에서 읽어온 데이터를 메모리에 캐시하여, 디스크 I/O를 줄여 데이터베이스 성능을 향상`시킵니다. 데이터베이스의 모든 스토리지 엔진에서 사용될 수 있습니다. InnoDB 스토리지 엔진에서는 버퍼 풀이라고도 하는 InnoDB 캐시를 사용합니다. InnoDB 캐시는 데이터와 인덱스를 모두 캐시하며, innodb_buffer_pool_size 설정 값에 의해 크기를 조정할 수 있습니다.

따라서, `키 캐싱은 인덱스에 대한 캐싱 기법`으로, 메모리에 색인 데이터를 저장하여 `디스크 I/O를 줄이고`, 버퍼 풀은 모든 데이터와 인덱스를 메모리에 캐시하여 `디스크 I/O를 줄이는 기법`입니다.

**Reply by CODe5753** [reply]
개인적으로, 키 캐시와 버퍼 풀 기능을 아직 완전하게 이해하지 못해서 혼동되는 부분이 많은 것 같네요.
자세한 설명 감사합니다 지피티씨!

**Comment by menthamin** [comment]
인덱스 데이터를 빠르게 참조하려고한다.
인덱스의 데이터 덩어리를 메모리에 캐싱 해두고 사용한다.

vs. 이노디비: 데이터 버퍼풀?
- 데이터 버퍼풀: 인덱스와 데이터가 포함된 데이터 블락?을 저장한다?

각 엔진에서 캐싱을 위한 장소가 각각 어떻게 불리는지

**Comment by CODe5753** [comment]
4.2.7절에서 버퍼 풀에 대한 자세한 지식을 습득 후 한번 더 논의되면 좋은 내용인 것 같습니다.
따라서 해당 Discussion은 종료합니다.
